### ISSUE TITLE
Proxy: Add command line argument to optionally send a certificate to the backend server for authentication

### ISSUE THREAD

**Issue by Hans-Giebenrath** [improbable-eng/grpc-web#170]
The grpcwebproxy seems to be incompatible with a grpc server (C++ in my case) that is configured with GRPC_SSL_REQUEST_AND_REQUIRE_CLIENT_CERTIFICATE_AND_VERIFY.
*I think* to make that work, in backend.go an optional argument has to be added which takes a path to a certificate, which then has to be added to tlsConfig.Certificates.
Would it be possible to add such an option or are there objections against it?

**Comment by johanbrandhorst** [comment]
Thanks for opening this issue. This might be possible with a bit of work, as [apparently browsers support client certificates](https://serverfault.com/questions/312617/are-ssl-client-certificates-well-supported-by-all-major-browsers).

I think this would mean two different flags:
1. A new flag (or flags) that would support taking a client certificate PEM (cert and key) and using that to authenticate requests made to the backend.
1. Separately, a new flag that would take a CA cert PEM and require client certificate verification from the browser clients.

It _might_ be possible to merge these two into a single authentication step, but I think the problem will be how we can ensure the proxy is able to present the browser (client) certificate directly to the backend server without seeming like a MITM attack.

Interesting suggestion!

**Comment by johanbrandhorst** [comment]
I just re-read your suggestion and I think your intention was for option 1 above, but please note that this opens a security hole in your architecture, as anyone able to connect to your proxy will be able to masquerade themselves as an authenticated user.

**Comment by Hans-Giebenrath** [comment]
Yes, my suggestion is your option 1. You are correct with the security hole - in our system, we have additional authentication layers, hence this is not a problem for us specifically. Our concern is to authenticate the gRPC proxy which connects to the backend server, so no arbitrary gRPC proxy can connect. As an alternative right now, we only allow local gRPC proxies to connect to the backend, which is ok, too.
